### ISSUE TITLE
Printing or downloading pdf not working

### ISSUE THREAD

**Issue by OlegVekhov** [CollaboraOnline/online#2041]
Hello! Neither print command (opens empty window) nor download as pdf command (nothing happens when choosing) works.

All other options (download as doc, xls, odf, csv) works fast and perfect. Don't know where to check next.

I'm using Collabora installed on separate ubuntu server from native repositories, and NextCloud 21.

LOOLWSD
6.4.7 (git hash: 653c70f)
Работает на: 3b2c4b47
LOKit
Collabora Office 6.4-34 (git hash: ee7a230)
"Ubuntu 18.04.5 LTS"

**Comment by Thatoo** [comment]
Hello,

I guess we have the same issue here : https://github.com/CollaboraOnline/richdocumentscode/issues/133 : 

We have Nextcloud 21.0.1 and the laste version of CODE installed behind a reverse proxy nginx with the "2. SSL terminates at the proxy:" file from https://www.collaboraoffice.com/code/nginx-reverse-proxy/ .

It is working pretty well but we can't download pdf version from "file/download/pdf" . We can download all other format but pdf. If I check in the log file, I can found this kind of error every time I try to download a pdf :

    [ websrv_poll ] ERR Download file [/opt/lool/child-roots/77Eog0oZ8r67Mntj/tmp/user/docs/iqimcpExcN6WxyC30X53WCFeceoXN24QBnZ6725LAQrEKzztqmXOtXGY8EXdc4W9/NAMEOFMYFILE.pdf] not found.| wsd/LOOLWSD.cpp:3243

Also, I was wondering if it is possible to save as pdf (or export as pdf) within nextcloud? I don't need to download in my computer...

**Comment by Thatoo** [comment]
@mrkara : can I do something to help confirming this issue and highlight where does the problem come from?

**Comment by kingul** [comment]
Also not working for me, with Apache2 and NextCloud 22

**Comment by verymilan** [comment]
Same problem here, with nginx. I can't find logs about this somehow

**Comment by fluboi** [comment]
Running the container as privileged is indeed a working workaround ! 
Might be linked with https://github.com/CollaboraOnline/online/issues/779

**Comment by Thatoo** [comment]
> Same problem here, with nginx. It cannot find the PDF file for downloading (while other formats work). The parent directoy exists but it is empty... I believe it could be related to the known mount errors in LXC?
> 
> Edit: yes, after moving it to the host, it is fixed for me, so this is indeed a problem with lxc.

Could you explain what you did to solve the problem?
I have tried to clone my unprivileged CT to a priviledged CT in Proxmox, collabora is just not working at all.

**Comment by verymilan** [comment]
@Thatoo i am sorry but maybe you have a different problem on top of that one... Collabora worked pretty fine for me despite the mount errors. It only today revealed itself, that PDF downloads would break which was not acceptable. As a temporaly measure (and for debugging the problem), i have installed it from their repo directly on the Debian host and use it with the same reverse-proxy.

**Comment by Thatoo** [comment]
Sorry, I meant that with an unprivileged CT, (into proxmox) collabora online is working great except this printing/pdf issue but if I transform the CT into a privileged one, it doesn't work at all.

**Comment by Thatoo** [comment]
I have update Collabora Online Development Edition (CODE) to
LOOLWSD 6.4.10 (git hash: b4fa48ef)
Collabora Office 6.4-45 (git hash: a21347f)

I forgot to mention it is on a "Debian GNU/Linux 10 (buster)" CT of Proxmox with pve-manager/6.4-13/9f411e79 (running kernel: 5.4.106-1-pve)

The problem remain...

**Comment by Thatoo** [comment]
Could be related to https://github.com/CollaboraOnline/richdocumentscode/issues/72
but I have tried to add "nesting=1" to /etc/pve/lxc/xxx.conf without success...

**Comment by Thatoo** [comment]
I have installed Collabora Online with the same settings on a VM in Proxmox instead of an unprivileged CT and it does work perfectly!
So I guess the problem does come from proxmox unprivileged CT...

**Comment by Thatoo** [comment]
The solution was not "nesting=1" but "mknod=1"

I did add 
`features: mknod=1`
into the file /etc/pve/lxc/xxx.conf corresponding to the proxmox CT into proxmox host and I rebooted the CT and now I can download as PDF and print.
Thank you @fabian from Proxmox forum : https://forum.proxmox.com/threads/collabora-online-development-edition-in-unprivileged-ct-of-proxmox.93882/

**Comment by kingul** [comment]
@Thatoo 
That was your solution. What about us, with native installed CODE?

**Comment by Thatoo** [comment]
Indeed, and? You seem not happy... Was I wrong sharing this solution on this Thread?

**Comment by tbelway** [comment]
I'm experiencing a similar issue with print and pdf options, instead of a blank window just nothing happens. Console indicates 404 and the collabora logs indicate the file not found.

My instance is containerized though, not native install.

**Comment by daufinsyd** [comment]
Hi there, I'm also experiencing the issue using containerized version.
PDF is not working, other are working.
```
wsd-00001-00035 2021-11-26 14:03:53.284232 [ websrv_poll ] ERR  Download file [/opt/lool/child-roots/OpUUeXz6cCgKG6f2/tmp/user/docs/MeFf6A7MXwVHU4nv1McZZEFvpy0hm7hH5NhBOoW7f5xbhYbZUAv9BKm2z8pyDu1A/Murlius & Klopaçus.pdf] not found.| wsd/LOOLWSD.cpp:3388
```

Code is running as container (sha: 30827c1421b8).
I guess this issue can be confirmed ?

**Comment by kingul** [comment]
New log in coolwsd:

`
[ websrv_poll ] ERR  #18: attempted to remove: 515 which is > size: 0 clamped to 0| ./net/Socket.hpp:1210

[ websrv_poll ] ERR  #19: attempted to remove: 503 which is > size: 0 clamped to 0| ./net/Socket.hpp:1210

[ kitbroker_002 ] ERR  SaveAs Failed for id=export [/tmp/user/docs/HBG4Q9rxWwk4ImV5BNQbomS6TEBMVApUOK7ZrbseCAOARgUb0hYtS5e8BKwDCHjS/myfile.pdf]. error= exception: SfxBaseModel::impl_store <file:///tmp/user/docs/HBG4Q9rxWwk4ImV5BNQbomS6TEBMVApUOK7ZrbseCAOARgUb0hYtS5e8BKwDCHjS/myfile.pdf> failed: 0xc10(Error Area:Io Class:Write Code:16) /home/collabora/jenkins/workspace/core-cp-21.06-for-arm64-PI/sfx2/source/doc/sfxbasemodel.cxx:3153 /home/collabora/jenkins/workspace/core-cp-21.06-for-arm64-PI/sfx2/source/doc/sfxbasemodel.cxx:1735| kit/ChildSession.cpp:1122

[ kitbroker_002 ] ERR  error: cmd=downloadas kind=saveasfailed| ./common/Session.hpp:139

[ docbroker_002 ] ERR  downloadas error failure: saveasfailed| wsd/ClientSession.cpp:1497`

**Comment by Kevin8675** [comment]
I still have this issue with Nextcloud 23, RichDocuments 5.0.1, and CODE version 21.11.0.6. Whenever I try to print or export as a pdf, I get this error:
"Document cannot be exported. Please contact storage server administrator"

Looking at the collabora logs I see this when I try to print:
```
nextcloud-collabora | kit-00261-00031 2021-12-28 16:16:22.868512 +0000 [ kitbroker_00a ] ERR  SaveAs Failed for id=print [/tmp/user/docs/rUoiAM4Qpg5sw3DZ3Q5sLgOh6HtyyxNdEK5FkJDdLF5LyNqQPtRBBfRpQ8m1Yk1U/print.pdf]. error= exception: SfxBaseModel::impl_store <file:///tmp/user/docs/rUoiAM4Qpg5sw3DZ3Q5sLgOh6HtyyxNdEK5FkJDdLF5LyNqQPtRBBfRpQ8m1Yk1U/print.pdf> failed: 0xc10(Error Area:Io Class:Write Code:16) /home/collabora/jenkins/workspace/core-cp-21.06-for-arm64-PI/sfx2/source/doc/sfxbasemodel.cxx:3153 /home/collabora/jenkins/workspace/core-cp-21.06-for-arm64-PI/sfx2/source/doc/sfxbasemodel.cxx:1735| kit/ChildSession.cpp:1122
nextcloud-collabora | kit-00261-00031 2021-12-28 16:16:22.868685 +0000 [ kitbroker_00a ] ERR  error: cmd=downloadas kind=saveasfailed| ./common/Session.hpp:139
nextcloud-collabora | wsd-00001-00282 2021-12-28 16:16:22.868890 +0000 [ docbroker_00a ] ERR  downloadas error failure: saveasfailed| wsd/ClientSession.cpp:1497
```
And when I try to export as pdf:
```
nextcloud-collabora | kit-00261-00031 2021-12-28 16:18:57.288167 +0000 [ kitbroker_00a ] ERR  SaveAs Failed for id=export [/tmp/user/docs/ysYWmuzJWZJkc3g2I7qDGCteIuh1OWfFiSiYf93woDlt9wQhwip1IplOGms0mmQX/test%20print%20document.pdf]. error= exception: SfxBaseModel::impl_store <file:///tmp/user/docs/ysYWmuzJWZJkc3g2I7qDGCteIuh1OWfFiSiYf93woDlt9wQhwip1IplOGms0mmQX/test%20print%20document.pdf> failed: 0xc10(Error Area:Io Class:Write Code:16) /home/collabora/jenkins/workspace/core-cp-21.06-for-arm64-PI/sfx2/source/doc/sfxbasemodel.cxx:3153 /home/collabora/jenkins/workspace/core-cp-21.06-for-arm64-PI/sfx2/source/doc/sfxbasemodel.cxx:1735| kit/ChildSession.cpp:1122
nextcloud-collabora | kit-00261-00031 2021-12-28 16:18:57.288315 +0000 [ kitbroker_00a ] ERR  error: cmd=downloadas kind=saveasfailed| ./common/Session.hpp:139
nextcloud-collabora | wsd-00001-00282 2021-12-28 16:18:57.288516 +0000 [ docbroker_00a ] ERR  downloadas error failure: saveasfailed| wsd/ClientSession.cpp:1497
```
I am running collabora in a docker container with docker compose. Here is my compose file:
```
  collabora:
    image: collabora/code:latest-arm64
    container_name: nextcloud-collabora
    restart: always
    ports:
      - 9980:9980
    environment:
      - domain=cloud\\.example\\.com
      - extra_params=--o:ssl.enable=false --o:ssl.termination=true
    cap_add:
      - MKNOD
    volumes:
      - /home/ubuntu/collabora/coolwsd.xml:/etc/coolwsd/coolwsd.xml
    privileged: true
```
If anyone could help I would really appreciate it.

**Comment by Moltey** [comment]
I confirm the bug.

Running as unprivileged container in podman and accessed via nextcloud.

COOLWSD version: 21.11.1.0

**Comment by krzysztofgal** [comment]
For me solution on proxmox lxc was:

1. Use debian package install instead of docker container,
2. Set `mknod=1` on container.

**Comment by jodab** [comment]
I can confirm this as well, running as unprivileged container in podman

**Comment by schefi** [comment]
Coolwsd creates an isolated rootfs for all the running document editor instances based on the systemplate. This root also has a dev directory, just like your normal Linux filesystem. For some reason pdf export and printing (which is pdf export under the hood) needs /dev/random and /dev/urandom for some reason algorithmic reason to be present in this isolated environment. But for obvious reasons, it can't create device nodes inside the container with mknod if the container is unprivileged. LXC creates those in standard locations before starting the container so this is not an issue most of the time. But there are some safe nodes that can be created even in unprivileged containers and lxc is prepared for this via the syscall intercept feature.
One solution would maybe to modify coolwsd in a way that it symlinks instead of mknod.

A possible workaround is to tick the mknod feature of proxmox, as others mentioned it before, or in case you are running plain LXD, like me, try this command:

`lxc config set container_name_goes_here security.syscalls.intercept.mknod true`
`lxc restart container_name_goes_here`

This way coolwsd's PDF export and printing will work within an unprivileged container as well. Just substitute your containers name in the command.

**Comment by Kevin8675** [comment]
> 

What about when running within docker? My container is privileged. You can look at the compose file above.

**Comment by microniko** [comment]
I had the same problem with a Proxmox container.
It resolved with adding `,mknod=1` at the end of the `features:` line in the container config file (`/etc/pve/nodes/yyyyyyy/lxc/xxxxx.conf`).

**Comment by kingul** [comment]
Problem fixed for me in 22.05.8.2. 
Coolwsd arm64.

**Comment by Ezinnem** [comment]
@kingul, Great to hear that =)
 
 @OlegVekhov , please can you test again with the recent 22.05.8.2 version?

**Comment by Moltey** [comment]
Still not working for me in a unprivileged / rootless container running on podman.

**Comment by Flexusma** [comment]
> The solution was not "nesting=1" but "mknod=1"
> 
> I did add `features: mknod=1` into the file /etc/pve/lxc/xxx.conf corresponding to the proxmox CT into proxmox host and I rebooted the CT and now I can download as PDF and print. Thank you @fabian from Proxmox forum : https://forum.proxmox.com/threads/collabora-online-development-edition-in-unprivileged-ct-of-proxmox.93882/

This did also fix it for me! Thanks a lot

**Comment by Thatoo** [comment]
Still need `features: mknod=1` even for version 22.05.8.4

**Comment by gitgt** [comment]
This issue and #133 had different cases. Some are resolved. I think that we should close this issue. 
I can't says why #5088 claims this one was for dockers, it started native but was mixed later. 

Remaining problem is of "Export to PDF2 for rootless containers. In that case error message is "The server encountered a failure error while parsing the exportas command."
That issue is known in Collabora docker, although not in progress, not sure that opening a new issue here would help.

**Comment by Thatoo** [comment]
> That issue is known in Collabora docker, although not in progress, not sure that opening a new issue here would help.

@gitgt Please, could you point us where is this issue for us to follow it?

**Comment by thebearon** [comment]
This should be fixed with CODE 23.05.1, to be released shortly.

**Comment by Traktormaster** [comment]
It is still not working, I've tried 23.05.1-2 and 23.05.2-2 built locally.

Download attempt logs:
```
kit-3867488-3867394 2023-07-25 16:50:34.374300 +0200 [ kitbroker_002 ] ERR  ToMaster-002: error: cmd=exportas kind=failure| common/Session.hpp:165
wsd-3867380-3867551 2023-07-25 16:50:34.374361 +0200 [ docbroker_002 ] ERR  ToClient-002: exportas error failure: failure| wsd/ClientSession.cpp:1623
```

Print attempt logs:
```
kit-3867488-3867394 2023-07-25 16:50:25.458001 +0200 [ kitbroker_002 ] ERR  ToMaster-002: SaveAs Failed for id=print [/tmp/user/docs/hX5PVYQkSQxXbMw1iy8E7s1rbJKKFGyKIAv2TrlC0eeyIxHHrxsPebbRvfb7Qeok/print.pdf]. error= exception: SfxBaseModel::impl_store <file:///tmp/user/docs/hX5PVYQkSQxXbMw1iy8E7s1rbJKKFGyKIAv2TrlC0eeyIxHHrxsPebbRvfb7Qeok/print.pdf> failed: 0xc10(Error Area:Io Class:Write Code:16) at /home/admin/tmp/libreoffice/sfx2/source/doc/sfxbasemodel.cxx:3221 at /home/admin/tmp/libreoffice/sfx2/source/doc/sfxbasemodel.cxx:1792| kit/ChildSession.cpp:1085
kit-3867488-3867394 2023-07-25 16:50:25.458061 +0200 [ kitbroker_002 ] ERR  ToMaster-002: error: cmd=downloadas kind=saveasfailed| common/Session.hpp:165
wsd-3867380-3867551 2023-07-25 16:50:25.458110 +0200 [ docbroker_002 ] ERR  ToClient-002: downloadas error failure: saveasfailed| wsd/ClientSession.cpp:1623
kit-3867488-3867394 2023-07-25 16:50:25.458213 +0200 [ kitbroker_002 ] ERR  ToMaster-002: error: cmd=exportas kind=failure| common/Session.hpp:165
wsd-3867380-3867551 2023-07-25 16:50:25.458249 +0200 [ docbroker_002 ] ERR  ToClient-002: exportas error failure: failure| wsd/ClientSession.cpp:1623
```

If I turn off the `capabilities` security feature in `coolwsd.xml`, then both work. I'm not running in any container (docker, LXC), so I'm not sure what to set the `mknod=1` on to try the workaround suggested previously.

**Comment by Thatoo** [comment]
Still need features: mknod=1 even for version 23.05.2.2
